File: sections/97_digit_broken_profile_dp/problems/b_forbidden_decimal_pattern/solution.py
import sys
from functools import lru_cache

def count(n, pattern):
    digits = list(map(int, str(n)))
    p = list(map(int, pattern))
    fail = [0] * len(p)
    for i in range(1, len(p)):
        j = fail[i - 1]
        while j and p[i] != p[j]:
            j = fail[j - 1]
        if p[i] == p[j]:
            j += 1
        fail[i] = j

    def step(j, d):
        while j and p[j] != d:
            j = fail[j - 1]
        if p[j] == d:
            j += 1
        return j

    @lru_cache(None)
    def dp(pos, j, tight, started):
        if pos == len(digits):
            return int(started or pattern != '0')
        limit = digits[pos] if tight else 9
        answer = 0
        for d in range(limit + 1):
            nt = tight and d == limit
            if not started and d == 0:
                answer += dp(pos + 1, 0, nt, False)
            else:
                nj = step(j, d)
                if nj < len(p):
                    answer += dp(pos + 1, nj, nt, True)
        return answer
    return dp(0, 0, True, False)

def main():
    data = sys.stdin.buffer.read().split()
    out = []
    for i in range(1, len(data), 2):
        out.append(str(count(int(data[i]), data[i + 1].decode())))
    print('\n'.join(out))
if __name__ == '__main__':
    main()

File: sections/97_digit_broken_profile_dp/problems/b_forbidden_decimal_pattern/solution.cpp
#include <bits/stdc++.h>
using namespace std;
using ll=long long;
string digits,pat;
vector<int>fail;
map<tuple<int,int,bool,bool>,ll>memo;
int step(int j,int d) {
    char c='0'+d;
    while(j&&pat[j]!=c)j=fail[j-1];
    if(pat[j]==c)j++;
    return j;
}
ll dfs(int pos,int j,bool tight,bool started) {
    if(pos==(int)digits.size())return started||pat!="0";
    auto key=tuple {
        pos,j,tight,started
    }
    ;
    if(!tight&&memo.count(key))return memo[key];
    int lim=tight?digits[pos]-'0':9;
    ll ans=0;
    for(int d=0;d<=lim;d++) {
        bool nt=tight&&d==lim;
        if(!started&&d==0)ans+=dfs(pos+1,0,nt,0);
        else {
            int nj=step(j,d);
            if(nj<(int)pat.size())ans+=dfs(pos+1,nj,nt,1);
        }
    }
    if(!tight)memo[key]=ans;
    return ans;
}
ll solve(ll n,string p) {
    digits=to_string(n);
    pat=p;
    fail.assign(p.size(),0);
    for(int i=1;i<(int)p.size();i++) {
        int j=fail[i-1];
        while(j&&p[i]!=p[j])j=fail[j-1];
        if(p[i]==p[j])j++;
        fail[i]=j;
    }
    memo.clear();
    return dfs(0,0,1,0);
}
int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    int q;
    cin>>q;
    while(q--) {
        ll n;
        string p;
        cin>>n>>p;
        cout<<solve(n,p)<<'\n';
    }
}
